[Kyh196201] 8주차 레스토랑 앱 꾸미기

kyh196201 opened the pull request:
트레이너님 안녕하세요 😀
벌써 마지막 과제네요 이번 주도 잘부탁드립니다!

첫 번째 강의 영상만 보고 `emotion.js`를 이용한 스타일링을 일부 진행했습니다.

---

현재까지 진행 상황입니다.

- [x] `emotion.js` 모듈 설치
- [x] HomePage 스타일링 및 테스트 수정
- [x] NavBar 컴포넌트 및 테스트 추가
- [x] AppHeader 컴포넌트 추가
- [x] Redux Toolkit 도입
- [x] 도메인 별로 slice 나누기(진행중)
- [ ] App 스타일 수정(일부 진행)

kyh196201 committed:
emotion 모듈 설치

kyh196201 committed:
@emotion/core 모듈 설치

kyh196201 committed:
index.html에 기본 스타일 추가

kyh196201 committed:
Header 컴포넌트 테스트 및 구현

App에 있는 header 영역을 분리하기 위해 Header 컴포넌트를 추가했습니다.

kyh196201 committed:
App 내부 header를 Header 컴포넌트로 대체

kyh196201 committed:
Header 컴포넌트 네이밍 수정

AppHeader 내부에서 Header 컴포넌트를 사용하기 위해서 네이밍을 AppHeader로 수정했습니다.

kyh196201 committed:
AppHeader 컴포넌트 스타일링

kyh196201 committed:
NavBar 컴포넌트 테스트 및 구현

HomePage에서 사용할 NavBar 컴포넌트를 구현했습니다.

kyh196201 committed:
HomePage 리팩토링

- HomePage에 타이틀, NavBar가 렌더링되었는지 확인하는 테스트 코드를 작성했습니다.
- HomePage 내부 ul 영역을 NavBar 컴포넌트로 대체했습니다.

kyh196201 committed:
NavBar 컴포넌트 스타일링

kyh196201 committed:
HomePage 스타일링

kyh196201 committed:
NavBar 스타일 수정

passwd10 commented on public/index.html:
```diff
@@ -3,6 +3,32 @@
+      a {
+        text-decoration: none;
+        color: inherit;
+      }
```
css 를 reset 하셨네요 👍 
추가로 `box-sizing` 속성도 정의해주시면 좋을 것 같아요!

kyh196201 replied:
감사합니다! 피드백 주신 부분 포함해서 추가 CSS 더 작성했습니다.😀

passwd10 commented on src/components/Container.jsx:
```diff
@@ -0,0 +1,8 @@
+
+const Container = styled.div({
+  fontSize: '1.6rem',
+  backgroundColor: '#fff',
```
color 는 6자리로 모두 표현하는게 좋아요

kyh196201 replied:
수정 완료했습니다!
그런데 6자리 모두 작성하는 특별한 이유가 있을까요?? 검색했는데 잘 안나와서요 😂

passwd10 replied:
디자이너와 협업시 3자리로 color 값을 정의한적이 거의 없는 것 같아요. 대부분의 color를 6자리로 표현하는데, 몇가지 color에 대해서만 임의로 3자리로 표현하게 된다면 나중에 해당 color를 검색해야할 때 한번에 찾기가 쉽지 않더라고요. 그 이후로 color 는 모두 6자리로 통일해서 표현하는 편이에요 ㅎㅎ

kyh196201 replied:
트레이너님 자세한 설명 감사합니다 :-) 실무적인 노하우가 있는 내용이었네요 !! 😎

passwd10 commented on src/components/home/NavBar.jsx:
```diff
@@ -0,0 +1,34 @@
+import { Link } from 'react-router-dom';
+
+export const ROUTES = [
```
라우터 관련된 데이터가 여러곳에서 사용되는거라면 NavBar 컴포넌트에서 관리하기보단, 데이터를 관리하는 폴더를 만들어서 모아놓으면 관심사를 분리할 수 있을 것 같아요!

passwd10 commented on src/components/home/NavBar.jsx:
```diff
@@ -19,16 +21,43 @@ export const ROUTES = [
+    marginRight: '0',
+  },
+
+  '& a': {
```
지금은 구조가 단순해서 문제가 없지만, 하위 태그가 많아질수록 태그에 의존한 스타일링 방식이 유지보수하기 어려워지더라고요 ㅎㅎ

kyh196201 replied:
피드백 감사합니다 트레이너님 🙂
말씀해주신 부분 관련해서 찾아보다가 공식문서에서 [예제](https://emotion.sh/docs/styled#targeting-another-emotion-component)를 찾았는데, 이렇게 적용하는 방법도 괜찮을까요?

passwd10 replied:
좋은 시도인 것 같아요! 적용해보시고 어떤점이 좋았는지, 불편했는지 정리해보시면 좋을 것 같아요 ㅎㅎ

passwd10 commented on src/App.jsx:
```diff
@@ -26,12 +29,9 @@ export default function App() {
-      </header>
+    <Container>
+      <AppHeader />
+
```
공백으로 구분하지 않아도 될 것 같아요!

kyh196201 replied:
수정했습니다 감사합니다!

passwd10 commented on src/components/Container.jsx:
```diff
@@ -0,0 +1,8 @@
+import styled from '@emotion/styled';
+
+const Container = styled.div({
```
CSS in JS 를 사용할 땐 스타일 컴포넌트만 따로 분리해서 사용하지 않아요. 이렇게 하면 css 파일을 분리해서 사용하는 것과 다를게 없어요! 왜 CSS in JS 을 사용하는지, 어떻게 사용되도록 설계 되었는지 찾아보시면 좋을 것 같아요 ㅎㅎ

만약 Container를 분리하고 싶다면 children 을 props로 받는 컴포넌트 형태로 만드는게 좋을 것 같아요

kyh196201 replied:
피드백 감사합니다!
정작 사용은 했는데, 왜 이걸 쓰는지를 깊게 생각해보지 않았네요 ..!
강의 영상에서 `CSS in JS` 관련해서 참고로 준 링크에 잘 나와있더라구요 감사합니다 🙂

passwd10 reviewed (commented):
마지막 주차까지 달려오느라 고생하셨어요 ㅎㅎ 끝까지 화이팅 해봐요! :)

kyh196201 committed:
Reset css 수정

kyh196201 committed:
App에서 Container 컴포넌트 직접 구현

- App에서 사용하는 Container 컴포넌트를 직접 구현하도록 수정했습니다.

kyh196201 committed:
reduxjs/toolkit 모듈 설치

kyh196201 committed:
reduxjs/toolkit을 이용해서 스토어 리팩토링

- application reducer를 관리하는 slice파일을 생성했습니다.
- actions.test.js에서 테스트하는 actions를 새로 추가한 slice.js에 있는 actions으로 변경했습니다.

kyh196201 committed:
slice 수정

reducer 내부에서 사용하는 action에 대한 action creator를 모두 export하도록 수정했습니다.

kyh196201 committed:
reducer 테스트 수정

- application slice에서 생성한 reducer를 테스트하도록 수정했습니다.
- 기존에 사용하던 action creator처럼 사용하기 위해서 application slice에서 생성한
action creator에 객체 형태로 파라미터를 전달하도록 수정했습니다.

kyh196201 committed:
App에서 setAccessToken 액션 수정

- setAccessToken을 추가한 application slice에 있는 액션으로 변경했습니다.

kyh196201 committed:
App 수정

- Container 컴포넌트의 css 색상 코드를 6자리로 수정했습니다.
- AppHeader와 Switch 사이의 불필요한 공백을 제거했습니다.

kyh196201 commented:
트레이너님 피드백 해주신 부분 수정하고, `reduxjs/toolkit` 도입해서 리팩토링 진행했습니다 🙂

`reducer, actions`에 대한 테스트는 변경한 slice 기준으로 수정완료했습니다.

다음 PR에는 각 컴포넌트에서 사용하는 action creator를 새로 추가한 slice 기준으로 변경할 예정입니다. 감사합니다.🙌

passwd10 commented on src/store/slice.js:
```diff
@@ -0,0 +1,230 @@
+  },
+};
+
+const { actions, reducer } = createSlice({
```
하나의 slice 에서 모든 상태를 관리하다보면 복잡도가 올라갈 것 같아요. 도메인별로 slice 를 분리해봐도 좋을 것 같아요!

kyh196201 replied:
트레이너님 피드백 감사합니다!
리덕스 공식문서에 있는 내용이랑 다른 분들 코드 참고하면서 진행하려고합니다. 🙂

passwd10 commented on src/store/slice.js:
```diff
@@ -0,0 +1,230 @@
+export function requestLogin() {
+  return async (dispatch, getState) => {
+    const { loginFields: { email, password } } = getState();
+
```
에러 핸들링도 시도해보시면 좋을 것 같네요!

kyh196201 replied:
앗 이 부분도 항상 고민하는 부분인데, 도메인 별로 슬라이스 분리하는 작업 후에 진행하도록 하겠습니다 감사합니다 ㅎㅎ

kyh196201 committed:
login 도메인에 해당하는 슬라이스 구현

- LoginPage에서 사용하는 상태를 관리하기 위해서 슬라이스와 테스트를 구현했습니다.

kyh196201 committed:
appSlice 추가

- 특정 도메인에 속하지 않는 accessToken같은  전역 상태를 관리하기 위해서 appSlice를 추가했습니다.

kyh196201 committed:
store 파일 위치 변경

kyh196201 committed:
App 컴포넌트 파일 위치 변경

kyh196201 committed:
store reducer에 rootReducer 세팅

- 각 도메인 별로 reducer를 분리하기 위해서 rootReducer를 구현했습니다.

kyh196201 committed:
loginSlice, appSlice selector 구현

- LoginFormContainer에서 accessToken과 loginFields의 값을 가져오기 위해
슬라이스 별로 selector를 구현했습니다.

kyh196201 committed:
LoginFormContainer email, password 입력하는 로직 변경

- store의 state가 변경되어서 loginFields와 accessToken을 가지고오는 selector를 변경했습니다.
- LoginPage, LoginFormContainer 내부에서도 변경된 state를 반영해서
useSelector의 구현을 변경했습니다.

passwd10 reviewed (commented):
> 1. accessToken의 위치
스타일 가이드에 따라서 도메인 별로 폴더를 나누었는데, accessToken과 같이 도메인에 종속(?)되어있지 않은 상태는 어디에 선언하면 좋을지 궁금합니다. 우선 appSlice를 만들어서 이곳에 선언했습니다!

잘 생각해보면 accessToken 을 이용하는 컴포넌트들은 결국 login 여부를 판단하기 위해 accessToken 을 사용하고 있기 대문에 login 이라는 도메인과 관련이 있다는 생각이 들어요. loginSlice 하위에 두어도 좋을 것 같아요! 

> 2. loginReducer.test.js의 파일 명
loginSlice.js로부터 import한 reducer를 테스트하기 위한 파일을 만들었는데, 파일 명이 적절한건지 궁금합니다.
loginSlice.test.js가 맞을까요??

action에 대한 테스트도 진행해야하니, loginSlice.test.js 내부에서 reducer 와 action을 모두 테스트하는건 어떠신가요?

> 3. 작업 진행 순서
먼저 각 도메인별로 slice와 테스트를 작성한 후 rootReducer로 모은다음 기존에 있는 configureStore({ reducer: })에 전달할 계획인데, 슬라이스를 구현하는대로 하나씩 추가하면서 리팩토링하는게 더 좋을까요??

처음부터 모든 상황을 고려해서 도메인을 나누기보다는 분리가 필요한 상황이 생길때마다 하나씩 리팩토링하는 방식으로 진행해보세요! 이런 점진적 설계방식을 `YAGNI(you aren't going to need it)` 라고 해요

https://becomereal.tistory.com/126

kyh196201 committed:
loginSlice에 대한 테스트 파일명 수정

kyh196201 committed:
accessToken을 loginSlice에서 관리하도록 수정

kyh196201 committed:
appSlice 제거

kyh196201 committed:
login 기능 구현

- loginSlice에 requestLogin 비동기 액션을 구현했습니다.

kyh196201 committed:
storage removeItem 함수 구현

- 로그아웃 시 로컬스토리지에 있는 acessToken을 제거하기 위해서 removeItem을 구현했습니다.

kyh196201 committed:
logout 기능 구현

- loginSlice에 logout, requestLogout 액션을 추가했습니다.
- logout 액션은 간단하게 accessToken을 클리어하는 액션이고,
requestLogout는 logout 액션을 dispatch하고,
로컬스토리지에 있는 accessToken을 제거하도록 구현했습니다.

kyh196201 committed:
새로고침해도 로그인 유지하도록 수정

- loginSlice에 있는 setAccessToken을 dispatch 하도록 수정했습니다.

kyh196201 committed:
regionsSlice 추가

kyh196201 committed:
categoriesSlice 추가

kyh196201 committed:
rootReducer에 regionsReducer, categoriesReducer 추가

kyh196201 committed:
RegionsContainer 리팩토링

- regionsSlice에 있는 state를 사용하도록 수정했습니다.
- 초기 regions가 없을 경우 조회 실패 메시지를 렌더링하도록 수정했습니다.

kyh196201 committed:
RestaurantsPage, App 테스트 수정

- useSelector에서 state의 구현을 regions slice에 맞게 변경했습니다.

kyh196201 commented:
트레이너님 안녕하세요!
마지막 과제인데 진행 속도가 너무 더디네요 ㅠㅠ
[답변 내용](https://github.com/CodeSoom/react-week8-assignment-1/pull/91#pullrequestreview-1054079889) 너무 감사합니다 😊

`reducer`를 여러 slice로 분리하는 과정에서 테스트랑 오류가 많이 발생해서 우선 `regionsSlice`, `categoriesSlice`만 추가했습니다.

--- 

질문이 있습니다.

`regions slice`에서 지역 목록 데이터를 관리하고 있는데, selector를 통해 접근할 때 `state.regions.regions` 이렇게 네이밍이 중복되더라구요.
이럴때 `state.regions.items`이렇게 변수명 변경하는 것도 괜찮을까요??

passwd10 commented on src/features/login/loginSlice.test.js:
```diff
@@ -1,14 +1,20 @@
 
 describe('loginSlice', () => {
-  // let store;
+  let store;
```
let 을 사용하지 않고 store 를 사용해보세요!

passwd10 commented on src/features/login/loginSlice.test.js:
```diff
@@ -59,6 +61,36 @@ describe('loginSlice', () => {
     });
   });
 
+  describe('logout', () => {
```
describe 로 한번 더 감싸서 reducer 인지 action인지 구분해주시면 테스트 파악이 더 수월해질 것 같아요!

passwd10 reviewed (commented):
> regions slice에서 지역 목록 데이터를 관리하고 있는데, selector를 통해 접근할 때 state.regions.regions 이렇게 네이밍이 중복되더라구요.
이럴때 state.regions.items이렇게 변수명 변경하는 것도 괜찮을까요??

컴포넌트에선 어차피 regionsSelector 로 상태를 가져오기 때문에 regions.regions 로 사용해도 크게 불편함을 느끼진 않을꺼에요. 편한대로 네이밍하시면 될 것 같아요!

메모이제이션을 이용해서 성능최적화 작업도 진행해보세요 :)

kyh196201 committed:
RegionsContainer 리팩토링

- 지역 선택 시 regionsSlice의 selectRegion을 dispatch하도록 수정했습니다.
- RegionsContainer가 렌더링될 때 지역 목록을 조회하도록 수정했습니다.

kyh196201 committed:
CategoriesContainer 리팩토링

- categories가 없을 경우에 조회 실패 메시지를 렌더링하도록 수정했습니다.
- categoriesSlice에 있는 categories, selectedCategory를 사용하도록 수정했습니다.

kyh196201 committed:
CategoriesContainer

- 지역 선택 시 categoriesSlice의 selectCategory을 dispatch하도록 수정했습니다.
- CategoriesContainer가 렌더링될 때 카테고리 목록을 조회하도록 수정했습니다.

kyh196201 committed:
변경된 store state에 맞게 RestaurantsPage, App 테스트 수정

kyh196201 committed:
restaurantsSlice 추가

kyh196201 committed:
rootReducer에 restaurantsReducer 추가

kyh196201 committed:
restaurants fixture 수정

- 레스토랑 목록 렌더링 테스트를 위해서 항목을 추가했습니다.

kyh196201 committed:
restaurantsSlice 테스트 수정

- restaurants fixture 수정하면서 발생한 오류를 수정했습니다.

kyh196201 committed:
restaurants fixture 두번째 레스토랑 id 수정

kyh196201 committed:
restaurantsSlice에 restaurantsSelector 추가

kyh196201 committed:
RestaurantsContainer 리팩토링

- restaurantsSlice에 있는 셀렉터를 사용하도록 수정했습니다.
- 레스토랑 목록이 없을 경우 조회 실패 메세지를 렌더링하도록 수정했습니다.

kyh196201 committed:
CategoriesContainer, RegionsContainer 리팩토링

- restaurantsSlice에 있는 loadRestaurants를 dispatch하도록 수정했습니다.

kyh196201 committed:
restaurantSlice 추가

kyh196201 committed:
restaurant fixture 수정

- 리뷰 목록 렌더링에 대한 테스트를 위해서 reviews를 추가했습니다.

kyh196201 committed:
rootReducer에 restaurantReducer 추가

kyh196201 committed:
RestaurantContainer 리팩토링

- restaurantSlice에 있는 selector를 사용하도록 수정했습니다.

kyh196201 committed:
RestaurantPage 테스트 수정

- restaurantReducer가 추가되어서 변경된 state에 맞게 useSelector의 구조를 변경했습니다.

kyh196201 committed:
RestaurantContainer 리팩토링

- accessToken selector 수정

kyh196201 committed:
RestaurantPage 테스트 오류 수정

kyh196201 committed:
RestaurantContainer 리팩토링

- restaurantSlice에 있는 loadRestaurant을 dispatch하도록 수정했습니다.

kyh196201 committed:
reviewSlice 추가

kyh196201 committed:
rootReducer에 reviewReducer 추가

kyh196201 committed:
RestaurantContainer 리팩토링

- reviewSlice의 reviewFieldsSelector를 이용해서 reviewFields를 가져오도록 수정했습니다.

kyh196201 committed:
RestaurantPage 테스트 수정

- reviewFieldsSelector 실행 과정에서 오류가 발생해서
useSelector에 reviewSlice에 대한 내용을 추가했습니다.

kyh196201 committed:
reviewSlice changeField 액션 오류 수정

kyh196201 committed:
RestaurantContainer 리팩토링

- change 이벤트 발생 시 reviewSlice의 changeField을 dispatch 하도록 수정했습니다.

kyh196201 committed:
restaurantSlice에 setReviews 액션 추가

kyh196201 committed:
restaurantSlice에 loadReviews 비동기 액션 추가

kyh196201 committed:
reviewSlice에 sendReview 비동기 액션 추가

kyh196201 committed:
RestaurantContainer 리팩토링

- submit 이벤트 발생 시 reviewSlice에 있는 sendReview를 dispatch하도록 수정했습니다.

passwd10 commented on src/features/restaurant/restaurantSlice.js:
```diff
@@ -0,0 +1,34 @@
+
+export function loadRestaurant({ restaurantId }) {
+  return async (dispatch) => {
+    dispatch(setRestaurant(null));
```
null 값을 넣어주는 이유가 있을까요?

kyh196201 replied:
레스토랑 정보를 새로 불러올 때 기존에 있는 값을 `null`로 변경해서 loading 메시지를 보여주려고 한 것입니다!
피드백 주셨는데, 답변이 너무 늦었네요 ㅠㅠ 감사합니다!

passwd10 commented on src/LoginFormContainer.jsx:
```diff
@@ -4,29 +4,30 @@ import LoginForm from './LoginForm';
+
+  const accessToken = useSelector(accessTokenSelector);
 
   const handleChange = ({ name, value }) => {
```
LoginForm 에 memo 를 사용하셨는데, props 로 넘기는 핸들러 함수들이 렌더링될 때 마다 새로 만들어지고 있네요. 이렇게되면 어떤 문제가 발생할까요?

kyh196201 replied:
트레이너님 아직 성능 최적화와 관련된 부분은 적용하지 못했습니다 ..!
제출 시간은 지났지만 개인적으로 공부하면서 과제 마무리를 하고 있습니다.
이 부분은 성능 최적화 부분 진행하면서 공부해보겠습니다 감사합니다!

kyh196201 committed:
RestaurantsPage 리팩토링

- loadInitialData 제거

kyh196201 committed:
actions, actions.test.js 파일 삭제

kyh196201 committed:
사용하지 않는 파일 삭제

kyh196201 committed:
HomePage 컴포넌트 위치 변경

kyh196201 committed:
Header 타이틀 수정

kyh196201 committed:
ROUTES를 관리하는 파일 추가

- ROUTES가 여러곳에서 자주 사용되기 때문에 routes.js 파일로 분리했습니다.

kyh196201 committed:
HomePage 리팩토링

- NavBar 컴포넌트를 제거했습니다.

kyh196201 committed:
NavBar 컴포넌트 위치 변경

kyh196201 committed:
AppHeader 수정

- NavBar 컴포넌트를 추가했습니다.

kyh196201 committed:
ROUTES 순서 변경

kyh196201 committed:
HomePage 스타일 및 문구 추가

kyh196201 committed:
ROUTES home 페이지 경로 수정

kyh196201 committed:
App 전체 레이아웃 스타일 수정

- Switch 컴포넌트를 감싸는 Wrapper 컴포넌트를 만들어서 전체 여백을 조정했습니다.

kyh196201 committed:
NavBar 컴포넌트 스타일 수정

- 현재 경로와 일치하는 경우에 Navigation  컴포넌트에 특정 스타일을 추가했습니다.

kyh196201 committed:
파일 구조 변경

- login, restaurants 도메인에 해당하는 파일의 경로를 수정했습니다.

kyh196201 committed:
AboutPage 파일 위치 수정

kyh196201 committed:
NotFoundPage 파일 위치 수정

kyh196201 committed:
파일 위치 수정

- CategoriesContainer, RegionsContainer의 파일 위치를 수정했습니다.

kyh196201 committed:
파일 위치 수정

kyh196201 committed:
AboutPage 스타일 수정

kyh196201 committed:
e2e 테스트 수정

kyh196201 committed:
e2e 테스트 수정

kyh196201 committed:
RegionsContainer 스타일 수정

kyh196201 committed:
CategoriesContainer, RegionsContainer 스타일 수정

kyh196201 committed:
RestaurantsContainer 스타일 적용

kyh196201 committed:
컴포넌트 스타일링
